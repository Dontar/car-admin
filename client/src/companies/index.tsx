import {
    ChipField,
    ReferenceArrayField,
    Show, ShowProps, SimpleShowLayout, SingleFieldList, TextField} from "react-admin";

// export const CompanyList = (props: ListProps) => (
//     <List {...props}>
//         <Datagrid rowClick="edit">
//             {/* <TextField source="id" /> */}
//             <TextField source="client_name" />
//             {/* <TextField source="bulstat" /> */}
//             {/* <TextField source="representative" /> */}
//             {/* <TextField source="phone" /> */}
//             {/* <TextField source="mobile" /> */}
//             {/* <TextField source="region_name" /> */}
//             {/* <TextField source="municipality_name" /> */}
//             <TextField source="city" />
//             {/* <DateField source="postcode" /> */}
//             <TextField source="street_name" />
//             <TextField source="street_no" />
//             {/* <DateField source="blok" /> */}
//             {/* <TextField source="vhod" /> */}
//             {/* <DateField source="apartment" /> */}
//             {/* <DateField source="floor" /> */}
//             <ReferenceArrayField source="car_ids" reference="cars">
//                 <TextField source="dkn" />
//             </ReferenceArrayField>
//         </Datagrid>
//     </List>
// );

export const CompanyShow = (props: ShowProps) => (
    <Show {...props}>
        <SimpleShowLayout>
            <TextField source="id" />
            <TextField source="client_name" />
            <TextField source="bulstat" />
            <TextField source="representative" />
            <TextField source="phone" />
            <TextField source="mobile" />
            <TextField source="region_name" />
            <TextField source="municipality_name" />
            <TextField source="city" />
            <TextField source="postcode" />
            <TextField source="street_name" />
            <TextField source="street_no" />
            <TextField source="blok" />
            <TextField source="vhod" />
            <TextField source="apartment" />
            <TextField source="floor" />
            <ReferenceArrayField source="car_ids" reference="cars">
                <SingleFieldList linkType="show">
                    <ChipField source="dkn" />
                </SingleFieldList>
            </ReferenceArrayField>
        </SimpleShowLayout>
    </Show>
);

// export const CompanyEdit = (props: EditProps) => (
//     <Edit {...props}>
//         <SimpleForm>
//             <TextInput source="id" />
//             <TextInput source="client_name" />
//             <TextInput source="bulstat" />
//             <DateInput source="representative" />
//             <TextInput source="phone" />
//             <TextInput source="mobile" />
//             <TextInput source="region_name" />
//             <TextInput source="municipality_name" />
//             <TextInput source="city" />
//             <DateInput source="postcode" />
//             <TextInput source="street_name" />
//             <DateInput source="street_no" />
//             <DateInput source="blok" />
//             <DateInput source="vhod" />
//             <DateInput source="apartment" />
//             <DateInput source="floor" />
//             <ReferenceArrayInput source="car_ids" reference="cars">
//                 <TextInput source="id" />
//             </ReferenceArrayInput>
//         </SimpleForm>
//     </Edit>
// );
