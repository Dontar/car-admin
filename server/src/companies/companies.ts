import { getDB } from "../connection";
import rawSql, { in as $in, select } from 'sql-bricks-sqlite';
import { GetOneParams, GetOneResult, GetParams, GetResult, ICompany } from "../share/models";
import { isOne, prepareParams, stringIdsToArray } from "../share/utils";

export async function getCompanies(params: GetOneParams): Promise<GetOneResult<ICompany>>;
export async function getCompanies(params: Partial<GetParams>): Promise<GetResult<ICompany>>;
export async function getCompanies(params: Partial<GetParams> | GetOneParams): Promise<GetResult<ICompany> | GetOneResult<ICompany>> {
    const db = await getDB();

    let sql = select(
        'c.ID as id',
        'c.CLIENT_NAME as client_name',
        'c.EGN as bulstat',
        'c.REPRESENTATIVE as representative',
        'c.PHONE as phone',
        'c.MOBILE as mobile',
        'c.REGION_NAME as region_name',
        'c.MINUCIPALITY_NAME as municipality_name',
        'c.CITY as city',
        'c.POSTCODE as postcode',
        'c.STREET_NAME as street_name',
        'c.STREET_NO as street_no',
        'c.BLOK as blok',
        'c.VHOD as vhod',
        'c.APARTMENT as apartment',
        'c.FLOOR as floor',
        'group_concat(vc.VEHICLE_ID) as car_ids'
    ).from('clients c')
        .leftJoin('vc', { 'vc.CLIENT_ID': 'c.ID' })
        .where('c.CLIENT_TYPE', rawSql(2))
        .groupBy('c.ID');

    let total: number | undefined = 78689;
    if (isOne(params)) {
        sql.where({ 'c.id': params.id });
    } else {
        if (params.ids) {
            sql.where($in('c.id', ...params.ids));
        } else {
            sql = await prepareParams(params, sql);
        }
    }

    const { text, values } = sql.toParams();
    if (isOne(params)) {
        const row = await db.get<ICompany>(text, values);
        return {
            data: stringIdsToArray(row!)
        };
    } else {
        const rows = await db.all<ICompany[]>(text, values);
        return {
            data: stringIdsToArray(rows),
            total
        };
    }
}
