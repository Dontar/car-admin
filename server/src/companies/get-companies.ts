import { GetOneParams, GetParams } from '../share/models';
import { getSqlCount, isOne, processGetManyRefs, processPaginationAndSort } from '../share/utils';
import { createJsonStream } from '../share/json-stream';
import { sql, where } from '../share/sql-tag';

export function getCompanies(params: Partial<GetParams> | GetOneParams): NodeJS.ReadableStream {

    const qry = sql/* sql */`
        select
            c.id as id,
            c.client_name as client_name,
            c.egn as bulstat,
            c.representative as representative,
            c.phone as phone,
            c.mobile as mobile,
            c.region_name as region_name,
            c.minucipality_name as municipality_name,
            c.city as city,
            c.postcode as postcode,
            c.street_name as street_name,
            c.street_no as street_no,
            c.blok as blok,
            c.vhod as vhod,
            c.apartment as apartment,
            c.floor as "floor",
            json_group_array(vc.vehicle_id) as car_ids
        from clients c
        left join vc on	(vc.client_id = c.id)
        ${where('c.client_type = 2')}
        group by c.id
    `;

    let total = 78689;
    if (isOne(params)) {
        qry.where?.and({ 'id = ': params.id });
    } else {
        if (params.ids) {
            qry.where?.and('id in ', params.ids);
        } else {
            if (processGetManyRefs(params, qry)) {
                total = getSqlCount(qry);
            }
            processPaginationAndSort(params, qry);
        }
    }

    return qry.stream.pipe(createJsonStream(
        isOne(params)
            ? { prefix: '{ "data": ', postfix: '}' }
            : {
                prefix: '{ "data": [',
                postfix: `] ${!params.ids ? `, "total": ${total} ` : ''}}`
            }
    ));
}
