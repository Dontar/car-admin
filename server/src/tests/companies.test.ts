import { config } from 'dotenv';
import { getCompanies } from "../companies/companies";
config();

describe('Test companies data retrieval', () => {

    it('should get a company from db', async () => {
        const data = await getCompanies({
            id: 9351
        });
        expect(data).toBeDefined();
        expect(data).toMatchSnapshot();
    });

    it('should get 2 companies from db', async () => {
        const data = await getCompanies({
            ids: [9351, 9851]
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get companies from db with pagination', async () => {
        const data = await getCompanies({
            pagination: { page: 1, perPage: 2 }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get companies from db with sorting', async () => {
        const data = await getCompanies({
            pagination: { page: 1, perPage: 2 },
            sort: { field: 'id', order: 'ASC' }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });
});
