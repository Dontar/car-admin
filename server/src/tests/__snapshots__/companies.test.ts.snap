// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test companies data retrieval should get 2 companies from db 1`] = `
Object {
  "data": Array [
    Object {
      "apartment": "0",
      "blok": "0",
      "bulstat": "175153616",
      "car_ids": Array [
        6001,
        1687957,
        1863595,
      ],
      "city": "ЛОЗЕН",
      "client_name": "Бонев Транс ЕООД  ",
      "floor": "0",
      "id": 9351,
      "mobile": "0",
      "municipality_name": "Столична",
      "phone": "0",
      "postcode": "1151",
      "region_name": "София (столица)",
      "representative": "Генади Стефанов Бонев",
      "street_name": "ЯНКО ПАНАЙОТОВ",
      "street_no": "28",
      "vhod": "0",
    },
    Object {
      "apartment": "27",
      "blok": "1",
      "bulstat": "130988423",
      "car_ids": null,
      "city": "СОФИЯ",
      "client_name": "\\" Мечкуеви Ко \\" ООД  ",
      "floor": "2",
      "id": 9851,
      "mobile": "0",
      "municipality_name": "Столична",
      "phone": "0",
      "postcode": "1000",
      "region_name": "София (столица)",
      "representative": "",
      "street_name": "Незабравка",
      "street_no": "1",
      "vhod": "Б",
    },
  ],
  "total": 78689,
}
`;

exports[`Test companies data retrieval should get a company from db 1`] = `
Object {
  "data": Object {
    "apartment": "0",
    "blok": "0",
    "bulstat": "175153616",
    "car_ids": Array [
      6001,
      1687957,
      1863595,
    ],
    "city": "ЛОЗЕН",
    "client_name": "Бонев Транс ЕООД  ",
    "floor": "0",
    "id": 9351,
    "mobile": "0",
    "municipality_name": "Столична",
    "phone": "0",
    "postcode": "1151",
    "region_name": "София (столица)",
    "representative": "Генади Стефанов Бонев",
    "street_name": "ЯНКО ПАНАЙОТОВ",
    "street_no": "28",
    "vhod": "0",
  },
}
`;

exports[`Test companies data retrieval should get companies from db with pagination 1`] = `
Object {
  "data": Array [
    Object {
      "apartment": "0",
      "blok": "0",
      "bulstat": "131304899",
      "car_ids": Array [
        6003,
        6157,
        7368,
        534737,
        828833,
        829572,
        829929,
        1528401,
        1564742,
        1623274,
        1651107,
        1655869,
        1661547,
        1674812,
        1676459,
        1687766,
        1746879,
        1749209,
        1755759,
        1778061,
        1791273,
        1808684,
        1817960,
        1837693,
        1837696,
        1841973,
        1886048,
        1905375,
        1926652,
        2013426,
        2110795,
        2114331,
        2207729,
        2234344,
        2261575,
        2316075,
        2325492,
        2433843,
        2465044,
        2500953,
        2542927,
        2603590,
        2621869,
        2952511,
      ],
      "city": "СОФИЯ",
      "client_name": "СИ ЕФ ЕМ ООД  ",
      "floor": "0",
      "id": 9251,
      "mobile": "359889100001",
      "municipality_name": "Столична",
      "phone": "359889100001",
      "postcode": "1000",
      "region_name": "София (столица)",
      "representative": "1",
      "street_name": "УЛ. НИКОЛАЙ ХАЙТОВ БЛ. 2, ВХ. Д, ОФИСИ 8 И 9",
      "street_no": "0",
      "vhod": "0",
    },
    Object {
      "apartment": "0",
      "blok": "0",
      "bulstat": "175153616",
      "car_ids": Array [
        6001,
        1687957,
        1863595,
      ],
      "city": "ЛОЗЕН",
      "client_name": "Бонев Транс ЕООД  ",
      "floor": "0",
      "id": 9351,
      "mobile": "0",
      "municipality_name": "Столична",
      "phone": "0",
      "postcode": "1151",
      "region_name": "София (столица)",
      "representative": "Генади Стефанов Бонев",
      "street_name": "ЯНКО ПАНАЙОТОВ",
      "street_no": "28",
      "vhod": "0",
    },
  ],
  "total": 78689,
}
`;

exports[`Test companies data retrieval should get companies from db with sorting 1`] = `
Object {
  "data": Array [
    Object {
      "apartment": "0",
      "blok": "0",
      "bulstat": "131304899",
      "car_ids": Array [
        6003,
        6157,
        7368,
        534737,
        828833,
        829572,
        829929,
        1528401,
        1564742,
        1623274,
        1651107,
        1655869,
        1661547,
        1674812,
        1676459,
        1687766,
        1746879,
        1749209,
        1755759,
        1778061,
        1791273,
        1808684,
        1817960,
        1837693,
        1837696,
        1841973,
        1886048,
        1905375,
        1926652,
        2013426,
        2110795,
        2114331,
        2207729,
        2234344,
        2261575,
        2316075,
        2325492,
        2433843,
        2465044,
        2500953,
        2542927,
        2603590,
        2621869,
        2952511,
      ],
      "city": "СОФИЯ",
      "client_name": "СИ ЕФ ЕМ ООД  ",
      "floor": "0",
      "id": 9251,
      "mobile": "359889100001",
      "municipality_name": "Столична",
      "phone": "359889100001",
      "postcode": "1000",
      "region_name": "София (столица)",
      "representative": "1",
      "street_name": "УЛ. НИКОЛАЙ ХАЙТОВ БЛ. 2, ВХ. Д, ОФИСИ 8 И 9",
      "street_no": "0",
      "vhod": "0",
    },
    Object {
      "apartment": "0",
      "blok": "0",
      "bulstat": "175153616",
      "car_ids": Array [
        6001,
        1687957,
        1863595,
      ],
      "city": "ЛОЗЕН",
      "client_name": "Бонев Транс ЕООД  ",
      "floor": "0",
      "id": 9351,
      "mobile": "0",
      "municipality_name": "Столична",
      "phone": "0",
      "postcode": "1151",
      "region_name": "София (столица)",
      "representative": "Генади Стефанов Бонев",
      "street_name": "ЯНКО ПАНАЙОТОВ",
      "street_no": "28",
      "vhod": "0",
    },
  ],
  "total": 78689,
}
`;
